src/met.hxx

```diff
@@ -0,0 +1,413 @@
+        [](const ROOT::RVec<float> &genparticle_pt,
+           const ROOT::RVec<float> &genparticle_eta,
+           const ROOT::RVec<float> &genparticle_phi,
+           const ROOT::RVec<float> &genparticle_mass,
+           const ROOT::RVec<int> &genparticle_id,
+           const ROOT::RVec<int> &genparticle_status,
+           const ROOT::RVec<int> &genparticle_statusflag) {
+            ROOT::Math::PtEtaPhiMVector genparticle;
+            float genPx = 0.; // generator Z(W) px
+            float genPy = 0.; // generator Z(W) py
+            float visPx = 0.; // visible (generator) Z(W) px
+            float visPy = 0.; // visible (generator) Z(W) py
+            // now loop though all genparticles in the event
+            for (std::size_t index = 0; index < genparticle_id.size();
+                 ++index) {
+                // consider a genparticle,
+                // 1. if it is a lepton and fromHardProcessFinalState --> bit 8
+                // from statusflag and 1 from status
+                // 2. if it is isDirectHardProcessTauDecayProduct --> bit 10
+                // in statusflag
+                Logger::get("getGenMet")
+                    ->debug("Checking particle {} ", genparticle_id.at(index));
+                if ((abs(genparticle_id.at(index)) >= 11 &&
+                     abs(genparticle_id.at(index)) <= 16 &&
+                     (IntBits(genparticle_status.at(index)).test(8)) &&
+                     genparticle_status.at(index) == 1) ||
+                    (IntBits(genparticle_status.at(index)).test(10))) {
+                    Logger::get("getGenMet")->debug("Adding to gen p*");
+                    genparticle = ROOT::Math::PtEtaPhiMVector(
+                        genparticle_pt.at(index), genparticle_eta.at(index),
+                        genparticle_phi.at(index), genparticle_mass.at(index));
+                    genPx += genparticle.Px();
+                    genPy += genparticle.Py();
+                    // if the genparticle is no neutrino, we add the x and y
+                    // component to the visible generator component as well
+                    if (abs(genparticle_id.at(index)) != 12 &&
+                        abs(genparticle_id.at(index)) != 14 &&
+                        abs(genparticle_id.at(index)) != 16) {
+                        Logger::get("getGenMet")->debug("Adding to vis p*");
+                        visPx += genparticle.Px();
+                        visPy += genparticle.Py();
+                    }
+                }
+            }
+            ROOT::Math::PtEtaPhiMVector genBoson;
+            ROOT::Math::PtEtaPhiMVector visgenBoson;
+
+            genBoson.SetPxPyPzE(genPx, genPy, 0,
+                                std::sqrt(genPx * genPx + genPy * genPy));
+            visgenBoson.SetPxPyPzE(visPx, visPy, 0,
+                                   std::sqrt(visPx * visPx + visPy * visPy));
+            std::pair<ROOT::Math::PtEtaPhiMVector, ROOT::Math::PtEtaPhiMVector>
```

Comment:
can we build the complete 4-vector of the genboson including z-component? we need it to determine the mass for the z-pt-mass-reweighting

Reply:
added